### PR TITLE
Allow Symfony 7 versions of packages

### PR DESCRIPTION
**Pull request by alexander-schranz**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no 
| Fixed tickets | fixes # 
| Related issues/PRs | part of #7413 
| License | MIT
| Documentation PR | sulu/sulu-docs# 

### What's in this PR?

Allow Symfony 7 versions of packages.

### Why?

The hardest part in Sulu 1.6 -> 2.0 was that it was not only a Sulu upgrade but also a Symfony upgrade (3.4 -> 4.4). We should avoid this when we go from 2.x -> 3.0 so we should sure the last version of 2.x will be compatible with Symfony 7, this would make upgrades a lot easier as Sulu projects can first upgrade PHP, then Symfony and then jump to the next Sulu version.

#### Symfony 6.4 errors

 - [x] doctrine-bridge 6.3 with validator 7.0: `Testing PHP Fatal error:  Declaration of Symfony\Bridge\Doctrine\Validator\DoctrineInitializer::initialize(object $object) must be compatible with Symfony\Component\Validator\ObjectInitializerInterface::initialize(object $object): void in /home/runner/work/sulu/sulu/vendor/symfony/doctrine-bridge/Validator/DoctrineInitializer.php on line 34` ( https://github.com/symfony/symfony/pull/51630 )
 - [x] sulu SecurityBundle and security-http 7.0 `PHP Fatal error:  Declaration of Sulu\Bundle\SecurityBundle\Security\AuthenticationEntryPoint::start(Symfony\Component\HttpFoundation\Request $request, ?Symfony\Component\Security\Core\Exception\AuthenticationException $authException = null) must be compatible with Symfony\Component\Security\Http\EntryPoint\AuthenticationEntryPointInterface::start(Symfony\Component\HttpFoundation\Request $request, ?Symfony\Component\Security\Core\Exception\AuthenticationException $authException = null): Symfony\Component\HttpFoundation\Response in /home/runner/work/sulu/sulu/src/Sulu/Bundle/SecurityBundle/Security/AuthenticationEntryPoint.php on line 25` 
 - [x] doctrine/persistence 3.0 support
    - [x] Fix deprecations https://github.com/sulu/sulu/pull/7134
    - [x] Fix merge deprecations in fixtures https://github.com/sulu/sulu/issues/7184
    - [x] Allow doctrine/persistence 3.0 support https://github.com/sulu/sulu/pull/6643
    - [x] https://github.com/sulu/sulu/issues/7198
 - [x] Declaration of Sulu\Component\Content\Form\DataTransformer\DocumentToUuidTransformer::transform($document) must be compatible with Symfony\Component\Form\DataTransformerInterface::transform(mixed $value): mixed in /home/runner/work/sulu/sulu/src/Sulu/Component/Content/Form/DataTransformer/DocumentToUuidTransformer.php on line 32 https://github.com/sulu/sulu/pull/7190
 - [x] https://github.com/sulu/sulu/pull/7347
    - [x] VoterInterface changes
    - [x] UserProviderInterface changes
    - [x] Security Class changes
 - [x] https://github.com/sulu/sulu/pull/7357
    - [x] ContainerAwareInterface removal
 - [x] https://github.com/sulu/sulu/pull/7359
    - [x] LoaderInterface changes
    - [x] ExpressionLanguageProvider changes 
 - [x] https://github.com/sulu/sulu/pull/7371
    - [x] Add compatibility to Symfony 7 to ActivateResolveTargetEntityResolverPas
 - [x] https://github.com/sulu/sulu/pull/7368
    - [x] Add Symfony 7 compatibility to FieldDescriptorFactory
 - [x] https://github.com/sulu/sulu/pull/7369
    - [x] Fix PreviewKernel and RefreshCommand compatibility
 - [x] https://github.com/sulu/sulu/pull/7374
    - [x] Refactor doctrine.event_subscriber to doctrine.event_listener for Symfony 7 compatibility
 - [x] https://github.com/sulu/sulu/pull/7380
    - [x] Refactor JMS/Serializer annotations to attributes
 - [x] https://github.com/sulu/sulu/pull/7382
    - [x] Refactor ParameterBag in InputBag in Core Tests
 - [x] https://github.com/sulu/sulu/pull/7390
    - [x] Refactor HeaderBag to InputBag in Core Tests
 - [x] https://github.com/sulu/sulu/pull/7392
    - [x] Fix some return types in tests expections

#### Dependencies

Before we can test sulu against 7.0 we need to tackle the following external dependencies to be compatible with Symfony 7:

 - [x] friendsofsymfony/http-cache-bundle
    - [x] https://github.com/FriendsOfSymfony/FOSHttpCacheBundle/pull/613
    - [x] 3.x release
 - [x] friendsofsymfony/http-cache
    - [x] https://github.com/FriendsOfSymfony/FOSHttpCache/pull/552
    - [x] 3.x release
 - [x] dantleech/phpcr-migrations-bundle
    - [x] dantleech/phpcr-migrations (https://github.com/phpcr/phpcr-migrations/pull/18)
    - [x] requires doctrine/phpcr-bundle
    - [x] requires symfony-cmf/testing
 - [x] doctrine/phpcr-bundle
    - [x] https://github.com/doctrine/DoctrinePHPCRBundle/pull/387
 - [x] doctrine/doctrine-fixtures-bundle (Requires 4.0: https://github.com/doctrine/DoctrineFixturesBundle/pull/395)
    - [x] https://github.com/sulu/sulu/pull/7161
 - [x] friendsofsymfony/rest-bundle
    - [x] https://github.com/FriendsOfSymfony/FOSRestBundle/pull/2400
 - [x] friendsofsymfony/jsrouting-bundle
    - [x] https://github.com/FriendsOfSymfony/FOSJsRoutingBundle/pull/471
 - [x] handcraftedinthealps/rest-routing-bundle
    - [x] requires friendsofsymfony/rest-bundle
 - [x] jms/serializer-bundle
    - [x] https://github.com/schmittjoh/JMSSerializerBundle/pull/937
 - [x] massive/build-bundle
    - [x] requires matthiasnoback/symfony-dependency-injection-test Symfony 7 support
    - [x] https://github.com/massiveart/MassiveBuildBundle/pull/25
 - [x] massive/search-bundle
    - [x] requires matthiasnoback/symfony-dependency-injection-test Symfony 7 support
    - [x] https://github.com/massiveart/MassiveSearchBundle/pull/172
 - [x] symfony-cmf/routing-bundle
    - [x] https://github.com/symfony-cmf/routing-bundle/pull/488
 - [x] matthiasnoback/symfony-config-test https://github.com/SymfonyTest/SymfonyDependencyInjectionTest/releases/tag/5.0.0
 - [x] matthiasnoback/symfony-dependency-injection-test https://github.com/SymfonyTest/SymfonyDependencyInjectionTest/releases/tag/5.0.0
 - [x] phpcr/phpcr-shell https://github.com/phpcr/phpcr-shell/pull/224
     - [x] https://github.com/phpcr/phpcr-utils/pull/214 (release missing)
     - [x] https://github.com/phpspec/phpspec/pull/1452
     - [x] https://github.com/Behat/Behat/pull/1442
 - [x] scheb/2fa-bundle
     - [x] Allow 7.x version https://github.com/sulu/sulu/pull/7254
     - [x] own 7.x version: https://github.com/scheb/2fa/blob/7.x/composer.json
 - [x] toflar/psr6-symfony-http-cache-store
    - [x] https://github.com/Toflar/psr6-symfony-http-cache-store/pull/46
 - [x] stof/doctrine-extensions-bundle
    - [x] requires first gedmo/extension https://github.com/doctrine-extensions/DoctrineExtensions
 
#### Our projects Statuses:

https://github.com/sulu/sulu/issues/7413

### CODE REVIEW

File: src/Sulu/Bundle/SecurityBundle/Security/AuthenticationEntryPoint.php
Comment (alexander-schranz) [inline src/Sulu/Bundle/SecurityBundle/Security/AuthenticationEntryPoint.php]: BC Break as class not internal or final.

File: composer.json
Comment (alexander-schranz) [inline composer.json]: This fork should be removed as soon as merged and released.

File: src/Sulu/Bundle/SecurityBundle/Entity/User.php
Comment (alexander-schranz) [inline src/Sulu/Bundle/SecurityBundle/Entity/User.php]: @chirimoya should we cherry pick the bc breaks already into 2.6 branch. Think it would be better to make the bc breaks in the minor instead of a patch release?

Comment (chirimoya) [reply src/Sulu/Bundle/SecurityBundle/Entity/User.php]: 👍

**Comment by alexander-schranz** [comment]
Current issue with Symfony 7 is that `bin/console cache:clear` / `bin/websiteconsole cache:clear` fails.

The issue seems to be in the `symfony/doctrine-bridge` changes as downgrading to 6.4 does solve the issue. https://github.com/symfony/doctrine-bridge/compare/6.4...7.0

One problem is a problem on `ConsoleEvents::COMMAND` a profiler get instanced which has Twig in it with the MediaManager which has the TrashRepository in it:

Bildschirmfoto 2024-04-16 um 13 56 32

Which instanced at the end the `Repository` which seems not have the `EntityTargetResolverListener` instanced correctly not sure how that can happen. 

------

Another issue is the the Warmup of the StructureFormMetadataLoader same there but at the end the same issue:

Bildschirmfoto 2024-04-16 um 14 13 54

------

We sure could workaround with Lazy services here but we maybe should find a solution in Symfony itself to avoid that projects run into the same problem.

So we require to find out why the doctrine metadata is not instanced here correctly.

--------

**UPDATE**: Resolved via: https://github.com/sulu/sulu/pull/7371

File: composer.json
Comment (alexander-schranz) [inline composer.json]: TODO revert this after http cache component and bundle is released

File: composer.json
Comment (alexander-schranz) [inline composer.json:97]: Proxy manager bridge is discontinued for Symfony 7.0 we can remove it as soon as we not longer supports. The first Symfony verison not requiring Proxy Manager is 6.2

**Comment by alexander-schranz** [comment]
The next step to replace all `doctrine.event_subscriber` with `doctrine.event_listener`:

Bildschirmfoto 2024-04-17 um 09 49 41

**Comment by alexander-schranz** [comment]
Based on the errors I think we need to convert our JMS Serializer annotations to attributes but not 100% sure. /cc @TheCadien / @mamazu if you want to have a look at it let us know.

Looks like there is also rector rule todo the change: https://github.com/rectorphp/rector-symfony/blob/b8126e8bf6e239fd53329fcd715ea117a4b4c34c/config/sets/jms/annotations-to-attributes.php#L15-L47 

The only thing which we need to check if the attributes loader is also supported for Symfony 5.4 JMS Bundle.

PS: The other annotations like Rest Routing Loader should currently stay as we implemented a bc layer for them.

**Comment by alexander-schranz** [comment]
So next step would be replace in our tests the `ParameterBag` with the `InputBag`: 

Bildschirmfoto 2024-04-24 um 16 48 26

https://github.com/search?q=repo%3Asulu%2Fsulu%20%22-%3Equery%20%3D%20new%20ParameterBag%22&type=code

Think we can even target for this refactoring 2.5 branch.

**Comment by alexander-schranz** [comment]
Tests seems to run now. The next step is to wait until https://github.com/jackalope/jackalope-doctrine-dbal/pull/447 is merged. When that is merged we can merge https://github.com/sulu/skeleton/pull/245 and as a next part bring our skeleton to Symfony 7. Thx @mamazu, @TheCadien and all other contributers and package maintainers that did make this possible.

That is required that we can test on prod env the new versions of FOSHttpCache and FOSHttpCache `3.x@dev` branches and give @dbu feedback if all works like expected. If that works @dbu can tag the last required dependencies for Symfony 7 compatibility.

After that all dependencies are then compatible we can then merge this one and merge then the Skeleton update to Symfony 7.

The only thing we need to tackle to make `lowest` here green is to configure: 

```
security:
    enable_authenticator_manager: true
```

via a `security.php` dependently if we use Symfony 5.4. For Symfony 6 true is default and for Symfony 7 the config does not exists.

**Review by Prokyonn — APPROVED** [review]

**Pull request merged**
